File: lambda.hs
{-# LANGUAGE CPP #-} -- C preprocessor for OS detection with ifdef
module Lambda where
import Data.List (sort,nub,intercalate)
import Data.Function (on)

data Lambda = Var Int
            | Ap [Lambda]  -- application over many arguments: xyzw instead of (((xy)z)w)
            | L Int Lambda -- repeated abstraction: λ[u,v,w]... instead of λ[u]λ[v]λ[w]...
  deriving Eq -- alpha-equivalence is syntactic equivalence

-- "Smart" constructors, enforcing the internal lambda term structure invariants
-- It is _highly_ recommended to use these in all cases, except when entering a
-- lambda term "literal", such as the example combinators below.
_Ap :: [Lambda] -> Lambda
_Ap ((Ap ts):ts') = Ap (ts++ts')
_Ap ts = Ap ts
_L :: Int -> Lambda -> Lambda
_L k (L m t) = L (k+m) t
_L 0 t = t
_L k t = L k t

-- Generate a human-friendly variable name from a deBruijn index: u,v,w,x,y,z,u1,v1,w1,...
-- The names for bound & free variables are generated slightly differently, otherwise
-- we'd need to keep an explicit stream of the already generated names.
boundName, freeName :: Int -> String
boundName n = ("uvwxyz"!!(rem n 6)) : (if n < 6 then "" else (show (div n 6)))
freeName n = let (c:num) = boundName n in (c:'0':num)

-- Pretty-print a λ where supported - i.e. everywhere, except Windows
lambda :: String
#ifdef mingw32_HOST_OS
lambda = "lambda"
#else
lambda = "λ"
#endif

-- We keep an implicit context of names for both bound and free variables. In order
-- to distinguish between the two classes, at each step we only keep the number of
-- bound variables so far, and after determining whether a variable is bound or free,
-- we generate procedurally a name, determined by its deBruijn index.
instance Show Lambda where
  show = showCtx 0

-- A function also used during type inference, not to be called directly
showCtx :: Int -> Lambda -> String
showCtx n (Var i) = (if i < n then boundName $ n-1-i else freeName $ i-n)
showCtx n (Ap ts) = concatMap (showBr n) ts
  where showBr n t@(Var _) = showCtx n t -- does the same as showCtx, except put braces around complex terms (non-variables)
        showBr n t = "(" ++ showCtx n t ++ ")"
showCtx n (L k t) = lambda ++ "[" ++ (intercalate "," names) ++ "]" ++ showCtx (n+k) t
  where names = map boundName [n..n+k-1]

-- Pretty-print a term to make its structure more easily visible. To-do: DOT
pretty :: Lambda -> IO () -- disclaimer: not actually pretty
pretty = putStrLn . pretty' 0
  where pretty' :: Int -> Lambda -> String -- pretty-print at a given indentation level
        pretty' n (Var i) = replicate n ' ' ++ "Var " ++ (show i) ++ "\n"
        pretty' n (Ap ts) = replicate n ' ' ++ "Ap\n" ++ concatMap (pretty' (n+2)) ts
        pretty' n (L k t) = replicate n ' ' ++ "L " ++ (show k) ++ "\n" ++ pretty' (n+2) t

-- Some classic combinators.
i, k, ks, s, w, om, y, j :: Lambda
i  = L 1 (Var 0) -- I
k  = L 2 (Var 1) -- K
ks = L 2 (Var 0) -- K*
s  = L 3 (Ap [Var 2, Var 0, Ap [Var 1, Var 0]]) -- S
w  = L 1 (Ap [Var 0, Var 0])  -- omega 
om = Ap [w,w] -- the irreducible Omega term
y  = L 1 (Ap [L 1 (Ap [Var 1, Ap [Var 0, Var 0]]), L 1 (Ap [Var 1, Ap [Var 0, Var 0]])]) -- fixed-point Y-combinator
j  = L 1 (Ap [Var 0, s, k]) -- Chris Barker's iota combinator: jj = I, j(jj) = K*, j(j(jj)) = K, j(j(j(jj))) = S
        
-- Increase or decrease all free variables with k, knowing their indices are >= d
arrow :: Int -> Int -> Lambda -> Lambda
arrow k d (Var i) = Var $ if i >= d then i+k else i
arrow k d (Ap ts) = Ap $ map (arrow k d) ts
arrow k d (L n t) = L n $ arrow k (d+n) t -- with each new bound variable all other indices increase by 1

-- Substitution helper functions
up :: Int -> Lambda -> Lambda
up k = arrow k 0    -- bump all free variables with k, knowing their indices start from 0
down :: Lambda -> Lambda
down = arrow (-1) 1 -- reduce all free variables by 1, knowing their indices have been bumped to >=1

-- Nameless lambda substitution
subst :: Int -> Lambda -> Lambda -> Lambda
subst i n (Var j) = if i == j then n else (Var j)
subst i n (Ap ts) = _Ap $ map (subst i n) ts
subst i n (L k t) = _L k $ (subst (i+k) (up k n) t)

-- Substitution examples:
l1, l2 :: Lambda
l1 = L 1 (Ap [Var 0, L 1 (Ap [Var 1, Var 0, Var 3])]) -- λ[x]x(λ[y]xyz)
l2 = L 1 (Ap [Var 0, Var 1])                          -- λ[u]uv
-- subst 1 l2 l1 -> λ[x]x(λ[y]xy(λ[u]uv)) or something alpha-equivalent to it :)

-- Traverse the list in depth, looking to reduce only the
-- left-most regex, and return Nothing if no regex is found.
reduceLeftMost :: [Lambda] -> Maybe [Lambda]
reduceLeftMost [] = Nothing
reduceLeftMost (t:ts) = case betaStep t of Just t' -> Just (t':ts)
                                           Nothing -> (t:) <$> reduceLeftMost ts

-- Normal reduction strategy, corresponding to lazy evaluation.
betaStep :: Lambda -> Maybe Lambda
betaStep (Ap ((L k m):n:ns)) = Just result
  where subst' = down $ subst 0 (up 1 n) (_L (k-1) m)
        result = if null ns then subst' else _Ap (subst':ns)
betaStep (Ap ts) = _Ap <$> reduceLeftMost ts
betaStep (L k t) = _L k <$> betaStep t
betaStep t       = Nothing -- nothing to reduce, t is a variable

-- Reduction to beta-normal form, if such exists
-- (otherwise, return the same term, for simplicity)
beta :: Lambda -> Lambda
beta t = case betaStep t of Nothing -> t
                            Just t' -> if t == t' then t else beta t'

-- Beta-equivalence
(=~=) :: Lambda -> Lambda -> Bool
(=~=) = (==) `on` beta
--x =~= y = beta x == beta y

-- Check for lambda term structure correctness
valid :: Lambda -> Bool
valid (Var i) = i >= 0
valid (Ap (Ap _:_)) = False
valid (Ap ts) = not (null ts) && not (null $ tail ts) && all valid ts
valid (L _ (L _ _)) = False
valid (L k t) = k > 0 && valid t

-- List every step of the beta reduction process
betaSteps :: Lambda -> [Lambda]
betaSteps t
  | res == Nothing = [t]
  | t == t'        = [t]
  | otherwise      = t : betaSteps t'
  where res = betaStep t
        (Just t') = res -- lazily evaluated

-- If all we need is to display the steps
betaSteps_ :: Lambda -> IO ()
betaSteps_ = mapM_ print . betaSteps

File: Lambda.cpp
#include "Lambda.h"

Ap::Ap(std::initializer_list<Lambda> terms) : terms{} {
	// The terms are "compressed" - a term will contain all of its arguments in the same vector.
	// This means that ((xy)z) will be repesented as a single node (xyz), and it should be
	// impossible for the first term (the "function" applied) to be an application itself.
	if (const Ap* t0 = Lambda::get_if<Ap>(terms.begin()); t0) {
		this->terms = t0->terms;
		for (auto it = terms.begin() + 1; it != terms.end(); ++it) {
			this->terms.emplace_back(std::move(*it));
		}
	} else {
		this->terms = terms;
	}
	vassert(this->terms.size() >= 2);
}

L::L(const unsigned k, Lambda&& body) : k{ k } {
	// Consecutive abstractions \x\y.(...) should also be compressed into a single node: \xy.(...)
	vassert(k > 0 && "There should be at least one bound variable!");
	if (L* term1 = Lambda::get_if<L>(&body); term1) {
		this->k += term1->k;
		this->body = std::move(term1->body);
	} else {
		this->body = std::make_unique<Lambda>(std::move(body));
	}
}

// Prints a generic name from the list: u,v,w,x,y,z,u0,v0,w0,...
void printGenericName(std::ostream& os, const unsigned idx) {
	os << char('u' + idx % 6);
	if (idx >= 6) {
		os << (idx / 6) - 1;
	}
}

// A context gives a mapping from [0;n) to arbitrary names. Each new bound variable is
// pushed to the front of the context, bumping the indices of the previously bound variables.
// An implicit context maps to the numbers in [0;n) in reverse order (because of this index bumping).
class ImplicitContext {
	unsigned n;
public:
	ImplicitContext() : n{ 0 } {}
	~ImplicitContext() {
		vassert(n == 0 && "Error: context should be empty after printing");
	}
	// Print the mapped name of a given variable.
	void printName(std::ostream& os, const unsigned idx) const {
		// The term may have free variables - for them the context has no mapping
		if (idx < n) {
			// We "flip" the index due to the reverse order of the numbers
			printGenericName(os, n - 1 - idx);
		} else {
			printGenericName(os, idx);
		}
	}
	// Push k new variables
	void push(const unsigned k) {
		n += k;
	}
	// Pop the k newest variables
	void pop(const unsigned k) {
		n -= k;
	}
	// Current size of the context
	unsigned size() const {
		return n;
	}
};

// The named context maps the first bound variables to given strings, but otherwise behaves as an
// implicit context. For named contexts pushing and popping expand and contract the visible "space"
// of names. Of course, once the given names run out, the implicit context behaviour will kick in.
class NamedContext : public ImplicitContext {
	const std::vector<std::string_view>& names;
public:
	NamedContext(const std::vector<std::string_view>& names) : ImplicitContext{}, names { names } {}
	// Print the mapped name of a given variable.
	void printName(std::ostream& os, const unsigned idx) const {
		// We remap only the outermost variables (the ones with the highest DeBruijn indices).
		// This means the highest indices should map in reverse to indices [0..] in the names array.
		const unsigned idx1 = size() - 1 - idx;
		if (idx1 < names.size()) {
			os << names[idx1];
		} else {
			ImplicitContext::printName(os, idx + unsigned(names.size()));
		}
	}
};

// Visitor for pretty-printing a lambda term to a given output stream,
// parameterized by the context that's used to generate names for variable indices.
template <class Context>
struct PrintVisitor {
	std::ostream& os;
	Context& ctx;

	PrintVisitor(std::ostream& os, Context& ctx) : os{ os }, ctx{ ctx } {}
	// The visiting functions:
	void operator()(const Var& l) {
		ctx.printName(os, l.idx);
	}
	void operator()(const Ap& l) {
		for (const auto& term : l.terms) {
			// Put brackets only around the complex terms
			const bool needsBr = !Lambda::get_if<Var>(&term);
			if (needsBr) { os << '('; }
			Lambda::visit(*this, term); // Use the same visitor, to preserve the context
			if (needsBr) { os << ')'; }
		}
	}
	void operator()(const L& l) {
		// Push the names of the new k bound variables into the context
		ctx.push(l.k);
		os << '\\';
		for (unsigned k = l.k; k > 0; --k) {
			ctx.printName(os, k - 1);
		}
		os << '.';
		Lambda::visit(*this, *l.body);
		// It is obligatory to pop the new names from the context afterwards.
		ctx.pop(l.k);
	}
};

std::ostream& operator<<(std::ostream& os, const Lambda& l) {
	ImplicitContext ctx{};
	Lambda::visit(PrintVisitor{ os,ctx }, l);
	return os;
}

void printWithCtx(std::ostream& os, const Lambda& l, const std::vector<std::string_view>& names) {
	NamedContext ctx{ names };
	Lambda::visit(PrintVisitor{ os,ctx }, l);
	os << '\n';
}

// Visitor class for alpha-equivalence. Used for traversing two lambda terms in
// parallel, checking whether they have the same structure and the same content.
struct EquivVisitor {
	bool operator()(const Var& l1, const Var& l2) const {
		return (l1.idx == l2.idx);
	}
	bool operator()(const Ap& l1, const Ap& l2) const {
		if (l1.terms.size() != l2.terms.size()) {
			return false;
		}
		for (size_t i = 0; i < l1.terms.size(); ++i) {
			if (!Lambda::visit(*this, l1.terms[i], l2.terms[i])) {
				return false;
			}
		}
		return true;
	}
	bool operator()(const L& l1, const L& l2) const {
		return (l1.k == l2.k && Lambda::visit(*this, *l1.body, *l2.body));
	}
	// Catch-all clause (much more verbose than [](auto x, auto y){ ... })
	// for when std::visit does not find a better-matching clause.
	template <typename V1, typename V2>
	bool operator()(const V1&, const V2&) const {
		return false;
	}
};

bool operator==(const Lambda& l1, const Lambda& l2) {
	return Lambda::visit(EquivVisitor{}, l1, l2);
}
bool operator!=(const Lambda& l1, const Lambda& l2) {
	return !(l1 == l2);
}

// Normal reduction strategy, corresponding to lazy evaluation.
bool Lambda::betaStep() {
	vassert(false && "Not implemented!");
	return false;
}

unsigned Lambda::betaReduce() {
	unsigned numRedexes = 0;
	// Each beta-step reduces one redex.
	while (betaStep()) {
		++numRedexes;
	}
	return numRedexes;
}

bool Lambda::isValid() const {
	struct ValidVisitor {
		bool operator()(const Var& l) {
			return true; // A variable is always structurally valid
		}
		bool operator()(const Ap& l) {
			if (l.terms.size() < 2 || Lambda::get_if<Ap>(&l.terms[0])) {
				return false;
			}
			for (const Lambda& term : l.terms) {
				if (!Lambda::visit(*this, term)) {
					return false;
				}
			}
			return true;
		}
		bool operator()(const L& l) {
			const Lambda& body = *l.body;
			return (l.k > 0 && !Lambda::get_if<L>(&body) && Lambda::visit(*this, body));
		}
	};
	return Lambda::visit(ValidVisitor{}, *this);
}
